### DISCUSSION TITLE
Nested alpine.data conflicts

### DISCUSSION THREAD

**Discussion by dmcgillen** [alpinejs/alpine#3507] [1. Help] [answered]
Hi there. I'm new to Alpine so I'm almost certainly doing something wrong, but I'm having trouble when adding x-data to an element that has children that contain x-data. To add context, I'm writing an HTMX app and using Alpine to help with various things, in this case form inputs. I'm also using Alpine.data for CSP reasons.

As an example, I'm using Alpine.data to make a currency input that has a hidden input containing the raw value that I want submitted in the form, and an input shown to the user that uses x-mask to show a nice, formatted value. This all works fine.

````
window.Alpine.data('currency', () => ({
    init() {
        const parameters = JSON.parse(this.$el.getAttribute("x-data-parameters"))
        this.rawValue = parameters.value
        this.symbol = parameters.symbol
        this.decimalPlaces = parameters.decimalPlaces
        this.displayValue = this.rawValue
    },

    decoratedInput: {
        ['@input']() {
            this.rawValue = this.$el.value.replaceAll(/[,$]/g, '')
        },

        ['x-mask:dynamic']() {
            return this.symbol + this.$money(this.$input, ".", ",", this.decimalPlaces)
        }
    }
}))
````

````
<span x-data="currency" x-data-parameters=“{‘value’:85, ‘symbol’: ‘$’, ‘decimalPlaces’:2}”>
  <input name=“currency-input1” type="hidden" x-model="rawValue" value="85">
  <input type="text" x-bind:value="displayValue" x-bind="decoratedInput">
</span>

<span x-data="currency" x-data-parameters=“{‘value’:105, ‘symbol’: ‘$’, ‘decimalPlaces’:2}”>
  <input name=“currency-input2” type="hidden" x-model="rawValue" value="105">
  <input type="text" x-bind:value="displayValue" x-bind="decoratedInput">
</span>
````

As I said, this works fine. I then wanted to use x-data with the containing `form` element for a completely unrelated reason, but I then find that all my currency inputs have the same value. This happens adding a completely minimal x-data to the `form` element:

````
window.Alpine.data("formState", () => ({
    init() {
        console.log('initialising form')
    }
}))
````

````
<form x-data="formState" ...>
    ....
</form>
````

For context, I want to use x-data with my form to check on change whether the form entries have changed, so I can tell the user they have unsaved changes.

Anyway, the way I understood it from the docs is that I could use these Alpine.data "objects" multiple times, but when contained in a nested one it seems that it doesn't work and they interfere with each other. Do they share state? Am I doing something wrong? Should this work?

**Comment by ekwoka** [comment]
im not exactly sure what's causing that, but definitely don't do this `x-data-parameters` thing you're doing.

You can just pass an object to the constructor function

```js

Alpine.data('currency', (parameters) => ({
    ...parameters,
   init() {
     this.displayValue = this.value
   },
  displayValue: null,
})
```
This spreads the object directly into the component, also use default values on the keys for the things you instead to use.
```html
<span x-data="currency({value: 85, symbol: `$’, decimalPlaces: 2 })”>
```

Start with that.

to talk about how they share, Children have access to the parents data by name (assuming their own component hasn't implemented that name). Basically a value would go up to the first that uses it.

**Reply by JJ-Atkinson** [reply]
We are using the CSP version of alpine, so the evaluator doesn't support `parameters`. `x-data-parameters` is just json data.

**Reply by ekwoka** [reply]
ah, interesting. I'm not sure if that would really work, since it's not valid JSON that you're putting in there.

There might be some possible issues with how errors are handled in the CSP version. I'd recommend checking which element is being identified by `this.$el` since it may be the wrong one.

**Reply by dmcgillen** [reply]
Just to be clear, it works perfectly fine until `x-data` is added to `<form>` (or any other parent of all the inputs). It's only at that point that things break down. With that done, if I log the values in the `init()` for `"currency"` everything looks good, but if I add a log to the `"x-mask:dynamic"` part of decoratedInput, `this.$input` comes out the same for all of them. Logging `this.$el.id` gives (correctly) the different id for each, so it is looking at the correct elements, but the values look to have been replaced so they all end up with whatever the last one was.

**Reply by ekwoka** [reply]
Quite curious. Did you build the csp version yourself from the current master branch?

I'll see if I can find some time to dig into what might be happening, cause from the code it certainly feels like it should work, but I haven't gone through the internals of mask.

I might check if this code works on the non-csp build first to see if that gives any clues to what might be the issue.

**Reply by dmcgillen** [reply]
Thanks for your help so far.

Just a couple of things I've discovered playing around this morning:
 - it still doesn't work if I use the regular build (and turn off our CSP checks) so I don't think it's related to the CSP build
 - It does seem to work if I convert it to using x-data, x-on etc directives instead of using Alpine.data

**Reply by ekwoka** [reply, marked as answer]
Do you have any of those values on keys in any of the parents?

You should try adding those keys directly to the input component, instead of only assigning them with `this.` in the init function.

I think it's causing the values to be added to the parent instead of the closest. To ensure it adds to the closest, you need to initialize them on the object, ideally with sensible defaults

**Reply by JJ-Atkinson** [reply]
I think we've got it. You mean like this?

```js
window.Alpine.data("somethingsomething", () => ({
    somethingsomething: 'default',
    init() {
        this.somethingsomething = getPropOff$el
    }
}))
```

Seems to work because of [how the proxy behaves](https://github.com/alpinejs/alpine/blob/68094c14144acbb13152ee5c1a5ee58c7231cd5f/packages/alpinejs/src/scope.js#L45-L101) and how the `x-data` directive works.

**Reply by ekwoka** [reply]
Yes absolutely!!!

This ensures that the component is fully aware of which keys the component should have and won't try to look for them or update them in a parent.

**Reply by dmcgillen** [reply]
Thanks very much for your help @ekwoka, makes total sense now - I hadn't realised what was actually happening under the hood.

**Reply by dmcgillen** [reply]
Just to add, it may be worth documenting this on the Alpine.data page of the docs? Thanks again.
